### PR TITLE
Refactor embeddings/rankings/tensorrtllm_engine converters to template-syle design

### PR DESCRIPTION
**Pull request by nv-hwoo**
This PR
- refactors the three converters into simpler template-style implementation
- adds unit tests for the resulting request body of the three converters (checks default case and case with request parameters)
- adds optional validation functionality to the `load_json_str` util function. This allows us to check/validate whether the json line contains the proper key name (e.g. `text_input` for LLM endpoints and `text` for embeddings or rankings).
- moved various input validation checks from converter to retriever side. It makes more sense for the retriever to check for these input related conditions, and it's good to place the checks closer to where the variables are defined.

### CODE REVIEW

File: genai-perf/genai_perf/inputs/converters/openai_embeddings_converter.py [hunk 1]
```diff
@@ -31,33 +31,25 @@
...
-        config: InputsConfig,
-    ) -> Dict:
-        pa_json: Dict[str, Any] = {"data": []}
 
-        for index, entry in enumerate(generic_dataset["rows"]):
-            iter_model_name = self._select_model_name(config, index)
-            payload = entry.get("payload", {})
-            input_values = payload.get("input")
+    def convert(self, generic_dataset: Dict, config: InputsConfig) -> Dict:
+        """
```
Comment (dyastremsky) [inline genai-perf/genai_perf/inputs/converters/openai_embeddings_converter.py:36]: Do we want to have these sorts of function descriptions in the base converter instead of copying them into each converter? Open either way on this, with a slight preference for just having the shared function descriptions in the base class.

File: genai-perf/genai_perf/parser.py [hunk 2]
```diff
@@ -135,6 +135,9 @@ def _check_conditional_args(
     Check for conditional args and raise an error if they are not set.
     """
 
+    if args.num_prompts <= 0:
```
Comment (dyastremsky) [inline genai-perf/genai_perf/parser.py:138]: Thanks for adding this check. Can you use ArgParse instead to make sure this arg is positive?

File: genai-perf/genai_perf/utils.py [hunk 3]
```diff
@@ -74,9 +74,19 @@ def load_json(filepath: Path) -> Dict[str, Any]:
         return load_json_str(content)
 
 
-def load_json_str(json_str: str) -> Dict[str, Any]:
+def load_json_str(json_str: str, func: Callable = lambda x: x) -> Dict[str, Any]:
+    """
+    Deserealizes JSON encoded string into Python object.
```
Comment (dyastremsky) [inline genai-perf/genai_perf/utils.py:79]:
```suggestion
    Deserializes JSON encoded string into Python object.
```

File: genai-perf/tests/test_rankings_converter.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
+# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (dyastremsky) [inline genai-perf/tests/test_rankings_converter.py:1]: If we're going to start making separate test files for each converter, we should see if we can have a subdirectory for the tests for the converters (same for the retriever tests). This can be in a separate PR, if necessary.

**Review by dyastremsky — COMMENTED** [review]
Great work! Appreciate all the refactoring and clean-up you did here.

File: genai-perf/tests/test_rankings_converter.py (hunk 4 above)
Comment (debermudez) [reply genai-perf/tests/test_rankings_converter.py:1]: I think we are going to have a few PRs soon that reorganize the file hierarchy.
I agree but i would prefer to have them be separate.

**Review by debermudez — COMMENTED** [review]
This is SO much more readable.
Fantastic work!

File: genai-perf/genai_perf/inputs/converters/openai_embeddings_converter.py (hunk 1 above)
Comment (nv-hwoo) [reply genai-perf/genai_perf/inputs/converters/openai_embeddings_converter.py]: removed the docstrings for the derived classes

File: genai-perf/genai_perf/parser.py (hunk 2 above)
Comment (nv-hwoo) [reply genai-perf/genai_perf/parser.py]: modified to use argparse for checking input values

File: genai-perf/genai_perf/utils.py (hunk 3 above)
Comment (nv-hwoo) [reply genai-perf/genai_perf/utils.py]: done

**Review by dyastremsky — APPROVED** [review]

File: genai-perf/tests/test_rankings_converter.py (hunk 4 above)
Comment (nv-hwoo) [reply genai-perf/tests/test_rankings_converter.py:1]: will address this in a separate PR
